File: Java/ReverseLinkedList.java
// Reverse a singly linked list.

public class ReverseLinkedList {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        while (head != null) {
            ListNode temp = head.next;
            head.next = prev;
            prev = head;
            head = temp;
        }
        return prev;
    }

    public ListNode reverseListUntil(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode n = reverseListUntil(head.next);
        head.next.next = head;
        head.next = null;
        return n;
    }
}

File: Python/reverse_linked_list.py
#!/usr/bin/env python

# Reverse a singly linked list.

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None or head.next is None:
            return head;
        n = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return n
